### PR TITLE
[WIP] MAHOUT-1991 - Add DBSCAN

### PR DESCRIPTION
**Pull request by AdityaAS**
DBSCAN - Google Summer of Code '17 - Apache Software Foundation (Apache Mahout)

Goals: Implement DBSCAN algorithm

Work done:
- Coded Sequential and Distributed versions of DBSCAN
- Added docs and unit tests
Work left to do:
- Improve distributed DBSCAN to make it more accurate
- Include cluster quality index calculation

Added the sequential version of the algorithm. Docs and Unit Tests where appropriate.
Will be adding rtree module and the approximate distributed algorithm soon.

Link to a short report describing the whole project can be found [here](https://docs.google.com/document/d/11sLJkwAftR2-Fmglju8c7dGTf1i3trkbEV2Zr9OlilU/edit?usp=sharing).

@rawkintrevo do help me out.

### Important ToDos
Please mark each with an "x"
- [x] A JIRA ticket exists (if not, please create this first)[https://issues.apache.org/jira/browse/ZEPPELIN/]
- [x] Title of PR is "MAHOUT-XXXX Brief Description of Changes" where XXXX is the JIRA number.
- [x] Created unit tests where appropriate
- [x] Added licenses correct on newly added files
- [x] Assigned JIRA to self
- [x] Added documentation in scala docs/java docs, and to website
- [x] Successfully built and ran all unit tests, verified that all tests pass locally.

If all of these things aren't complete, but you still feel it is
appropriate to open a PR, please add [WIP] after MAHOUT-XXXX before the
descriptions- e.g. "MAHOUT-XXXX [WIP] Description of Change"

Does this change break earlier versions?

Is this the beginning of a larger project for which a feature branch should be made?

### CODE REVIEW

**Comment by rawkintrevo** [comment]
Aditya and I talked about this on a hangout- for posterity: Needs to be refactored into incore/distributed, made to fit with the Mahout Algos framework, and distanceMetric should leverage mahout's current distance metrics.

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          }
+        }
+        if(data(curr, 3) == -1) {
+          data(curr, 3) = clusterId
+        }
+      }
+    }
+  }
+
+  def DBSCAN(data: DenseMatrix, Eps: Double, Minpts: Int): Unit = {
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: do we really care to force DenseMatrix vs. just generic Matrix?

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  def distanceMetric(arg1: DenseVector, arg2: DenseVector) : Double = {
+    var diffsqSum = -1.0
+    if(arg1.length != arg2.length) {
+      return diffsqSum
+    }
+    else {
+      val diff = arg1 - arg2
+      val diffsq = diff^2
+      diffsqSum = 0.0
+      for(i <- 4 until diffsq.length) {
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: DSL suggestion for this loop (not necessarily faster though, but DSL-er: ) ): 
    
    diffSqSum += diffsq(4 until diffsq.length).sum

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        diffsqSum += diffsq(i)
+      }
+      diffsqSum = math.sqrt(diffsqSum)
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: i'd just import scala.math._ and use just sqrt(a) etc.

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  Returns: A DenseVector containing the ID's of all the points that are at an Eps distance from point.
+   */
+  def findNeighbours(point: Int, data: DenseMatrix, Eps: Double) : DenseVector = {
+    val pointId: Int = data(point, 2).toInt
+    var neighbours: ArrayBuffer[Int] = new ArrayBuffer[Int]()
+    val pointData: DenseVector = dvec(data(pointId, ::))
+    var neighbourCount = 0
+    for(row <- data) {
+      if(row(0).toInt != pointId) {
+        val arg1 = dvec(row)
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: i am not sure if `dvec` use is warranted... row is already an `o.a.m.math.Vector`

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  /*
+  @args
+  point: Int - Id of the point whose neighbours are to be computed
+  data: DenseMatrix[Double] - The augmented matrix containing all the data points along with the 4 appended columns
+  Returns: A DenseVector containing the ID's of all the points that are at an Eps distance from point.
+   */
+  def findNeighbours(point: Int, data: DenseMatrix, Eps: Double) : DenseVector = {
+    val pointId: Int = data(point, 2).toInt
+    var neighbours: ArrayBuffer[Int] = new ArrayBuffer[Int]()
+    val pointData: DenseVector = dvec(data(pointId, ::))
```
Comment (dlyubimov) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: redundant dvec (imo), slice is already an `o.a.m.math.Vector`

**Comment by rawkintrevo** [comment]
@andrewpalumbo proposing merging this to feature branch.  It ended up being a bit hairier than we expected.

**Comment by andrewpalumbo** [comment]
Sounds good to me. Thx.



Sent from my Verizon Wireless 4G LTE smartphone


-------- Original message --------

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala (hunk 1 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: Nope. Changed all occurrences accordingly

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala (hunk 5 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: Updated

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala (hunk 4 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: Updated

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala (hunk 3 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN_Incore.scala]: Updated

File: website/docs/algorithms/clustering/dbscan/index.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+</div>
+
+
+### Example
+
+    //Using the InCoreDBSCAN module
+    val input = dense((1.0, 1.2, 1.3, 1.4), (1.1, 1.5, 2.5, 1.0), (6.0, 5.2, -5.2, 5.3), (7.0,6.0, 5.0, 5.0), (10.0, 1.0, 20.0, -10.0))
+
+    import org.apache.mahout.math.algorithms.clustering.InCoreDBSCAN
+    val distanceMetric = DistanceMetricSelector.select(DistanceMetricSelector.namedMetricLookup('Euclidean)))
```
Comment (rawkintrevo) [inline website/docs/algorithms/clustering/dbscan/index.md]: +1 utilizing existing distanceMetrics framework.

**Comment by rawkintrevo** [comment]
@AdityaAS @andrewpalumbo @dlyubimov 

Afaik, the R-Tree/Rect portion of this is solid- I could really use those for something else I'm working on- (ML/AI on Flink for upcoming talks).  

As RTrees area  core component for many hierarchical clustering algorithms, does anyone have any opinions to refactoring this JIRA into multiple- the first of which being "Add R-Tree implementation" (and then additional for DBSCAN/ etc.) 

Still proposing feature branch for most of this- just splitting the R-Trees which are done off into their own JIRA ticket/PR and merging. 

Thoughts?

**Comment by andrewpalumbo** [comment]
@rawkintrevo @AdityaAS @dlyubimov Admittedly, I've not had a chance to review this closely, I think that, If the RTrees are solid, moving them to something like `org.apache.mahout.maath.commmon` and committing separately would make sense.  We have other data structures from the pre-algo framework days that need to be shared as well, e.g.: `ConfusionMatrix`: https://github.com/apache/mahout/tree/master/math-scala/src/main/scala/org/apache/mahout/classifier/stats which could also go in such a package.

**Comment by andrewpalumbo** [comment]
@rawkintrevo to clarify, you're talking about a PR against `master`  for RTrees, correct?

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN.scala [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+      }, {
+        // Optionally Merge Clusters that are close enough
+        case (metadata1: Matrix, metadata2: Matrix) => {
+          // this does nothing- just returns the left matrix
+          metadata1
+        }
+      })
+
+    val model = new DBSCANModel(1)
+    model.summary = s"""foo the bar"""
```
Comment (andrewpalumbo) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN.scala]: @AdityaAS do you have a plan for the model summary?

File: math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala [hunk 8]
```diff
@@ -45,4 +48,17 @@ trait ClusteringSuiteBase extends DistributedMahoutSuite with Matchers {
     val epsilon = 1E-6
     (myAnswer.norm - correctAnswer.norm) should be <= epsilon
   }
+
+  test("inCore DBSCAN test") {
+//    import org.apache.mahout.math.algorithms.clustering._
+//
+//    val dm = DistanceMetricSelector.namedMetricLookup('Euclidean)
+////    val url: URL = getClass().getResource("./data21.txt")
+//    //TO DO //Need help
```
Comment (andrewpalumbo) [inline math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala]: @AdityaAS if you need some toy data files to for tests, you can place them in `$MAHOUT_HOME/examples/bin/resources".

File: math-scala/src/test/scala/org/apache/mahout/math/algorithms/data21.txt [hunk 9]
```diff
@@ -0,0 +1,23 @@
+21
```
Comment (andrewpalumbo) [inline math-scala/src/test/scala/org/apache/mahout/math/algorithms/data21.txt]: see above comment.

File: math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala (hunk 8 above)
Comment (AdityaAS) [reply math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala]: Ah. Thanks. I'll do that.

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    for(i <- 0 until n) {
+      squaredSum += dList(i)*dList(i)
+    }
+    sqrt(squaredSum)
+  }
+
+}
+
+//Represents a point with coordinates and n information
+case class Point(coordinates: List[Double]) extends Rectangle {
```
Comment (rawkintrevo) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala]: Why not move this to its own file?

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      squaredSum += (pt.bottomLeft(i) - bottomLeft(i))*(pt.bottomLeft(i) - bottomLeft(i))
+    }
+    squaredSum
+  }
+
+  override def wraps(rect: Rectangle): Boolean = false
+}
+
+//MBR is a physical manifestation of the Rectangle trait
+case class MBR(bottomLeft: List[Double], topRight: List[Double]) extends Rectangle {
```
Comment (rawkintrevo) [inline math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala]: same with case class and companion object. (move to own file)

File: math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala (hunk 8 above)
Comment (andrewpalumbo) [reply math-scala/src/test/scala/org/apache/mahout/math/algorithms/ClusteringSuiteBase.scala]: I apologize @AdityaAS I'd not noticed that this `.txt` file wasthis short-  No need to move to `$MAHOUT_HOME/examples/bin/resources` for Unit test input.  FYI though if you write an example for this to be placed in `examples`, that's where you would place `.txt` files or any other resources.

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN.scala (hunk 7 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/DBSCAN.scala]: Yes. I do. Will update it.

File: math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala (hunk 11 above)
Comment (AdityaAS) [reply math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala]: My idea was to keep classes describing the datastructures in one file and the logic in another. But if having different classes is food design I'll update it

**Comment by AdityaAS** [comment]
@rawkintrevo @andrewpalumbo having a separate PR for Rtrees is a good idea. RTrees can be used in other algorithms as well (kNN for example). Should I go ahead a create a new issue on JIRA?

**Comment by rawkintrevo** [comment]
@AdityaAS Yes please. Also don't forget to add a jira ticket.

**Comment by rawkintrevo** [comment]
status on this @AdityaAS ?

**Comment by rawkintrevo** [comment]
bump

**Pull request closed without merging**